syntax = "proto3";

message FuseChange {

  bytes uuid = 1;

  enum Operation {
    CHMOD = 0;
    CHOWN = 1;
    CREATE = 3;
    FSYNC = 4;
    LINK = 5;
    MKDIR = 6;
    OPEN = 7;
    RELEASE = 8;
    RMDIR = 9;
    SYMLINK = 10;
    TRUNCATE = 11;
    UNLINK = 12;
    UTIME = 13;
    WRITE = 14;
  }

  Operation op = 2;

  string path = 3;

  bytes data = 4;

  uint64 offset = 5;

  uint32 mode = 6;

  enum Flag {

    O_RDONLY    = 0;
    O_WRONLY    = 1;
    O_RDWR      = 2;

    O_APPEND    = 3;
    O_ASYNC     = 4;
    O_CLOEXEC   = 5;
    O_CREAT     = 6;
    O_DIRECT    = 7;
    O_DIRECTORY = 8;
    O_DSYNC     = 9;
    O_EXCL      = 10;
    O_LARGEFILE = 11;
    O_NOATIME   = 12;
    O_NOCTTY    = 13;
    O_NOFOLLOW  = 14;
    O_NONBLOCK  = 15;
    O_PATH      = 16;
    O_RSYNC     = 17;
    O_SYNC      = 18;
    O_TMPFILE   = 19;
    O_TRUNC     = 20;

  }

  repeated Flag flags = 7;

  uint32 uid = 8;
  uint32 gid = 9;

  bytes fh_uuid = 10;

  bool datasync = 11;

  string src = 12;

  uint32 length = 13;

  float atime = 14;
  float mtime = 15;

}
